sdk/dotnet/Pulumi.Automation/WorkspaceStack.cs

```diff
@@ -275,9 +275,25 @@ public Task RemoveAllConfigAsync(IEnumerable<string> keys, CancellationToken can
                 args.Add("--exec-kind");
                 args.Add(execKind);
 
-                var upResult = await this.RunCommandAsync(args, options?.OnStandardOutput, options?.OnStandardError, options?.OnEvent, cancellationToken).ConfigureAwait(false);
-                if (inlineHost != null && inlineHost.TryGetExceptionInfo(out var exceptionInfo))
-                    exceptionInfo.Throw();
+                CommandResult upResult;
+                try
+                {
+                    upResult = await this.RunCommandAsync(
+                        args,
+                        options?.OnStandardOutput,
+                        options?.OnStandardError,
+                        options?.OnEvent,
+                        cancellationToken).ConfigureAwait(false);
+                }
+                catch
+                {
+                    if (inlineHost != null && inlineHost.TryGetExceptionInfo(out var exceptionInfo))
```

Comment:
This is a little interesting here could you help me understand what is going on exactly? Presumably RunCommandAsync fails with an exception (catch), but we prefer to throw an exception captured by `inlineHost` if any. What is the original exception? It's not linked to or contain the inline host exception in any way?

Reply:
So now that the `LanguageRuntimeService` is returning a failure response correctly, that means that `RunCommandAsync` will throw `CommandException` if the caller's provided inline program throws an exception. But that's not the exception we want to throw to the caller - we want the actual inline exception to bubble to the caller, but since that is thrown on a different thread from the inline host inside the gRPC service we need to capture it separately so we can throw it here.

Reply:
This is to support the scenario described in the test `InlineProgramExceptionPropagatesToCaller`

Reply:
To clarify - this should not be a change in behavior for current consumers of inline programs because `InlineProgramExceptionPropagatesToCaller` was passing before this changeset. We just had to change the way we handle it internally because we were erroneously relying on the gRPC service always returning gracefully for inline programs

Reply:
Ah thanks for those pointers! Aha it makes a little more sense now.

Another curious thing here is this. Before the change, inlineHost.TryGetExceptionInfo check ran only if this.RunCommandAsync did not throw. After the change, it runs only if this.RunCommandAsync throws. 

Are we sure this is what we want? Should we maybe always run it?

```
try {

   try {
      upResult = await this.runCommandAsync()
   } catch (Exception e) {
      upExc = e;
   }
} finally {
  if (inlineHost != null &amp;&amp; inlineHost.TryGetExceptionInfo(out var exceptionInfo))                    
     exceptionInfo.Throw();
   }
}

if upExc != null {
   throw upExc;
}
```

Reply:
Technically every inline host exception should have caused `RunCommandAsync` to throw, but not every `RunCommandAsync` exception has an inline exception

Reply:
[This line here](https://github.com/pulumi/pulumi/pull/7299/commits/4c4beb84064d5f4f0d22f3e786c5b97332915046#diff-16d113734488cfbc4117f2542e438ae49970eb3b9ffcce5864e5fa96a1043638R64) is checking the same `ExceptionDispatchInfo` instance that `TryGetExceptionInfo` returns when true. So if it was not null, than we bailed in gRPC and should have a `CommandException` and `TryGetExceptionInfo` should be true